### ISSUE TITLE
[API] folder structure and taxonomy for optimization algorithms

### ISSUE THREAD

**Issue by fkiraly** [SimonBlanke/Hyperactive#126]
Following up on discussion from #121.

Mid-term, what should the folder structure for optimizations algorithm be? Including cases such as:

* the same algorithm coming from two packages (third party), potentially with different parameterization
* multiple algorithms from the same package, e.g., `gfo`

There are different taxonomies that one can apply here, two discussed were:

* semantic or algorithm type based, e.g., group multiple grid searches in one place
* dependency based, e.g., group all algos from `gfo` in one place

**Comment by fkiraly** [comment]
> How would it "mask" it?

If the current directory is `hyperactive.opt`, then `from sklearn import ...` will result in attempting to get it from that module, not the package `scikit-learn`. It is maybe an edge case, but it is general good practice to not mask popular packages or python base modules (e.g., that is why you would not like to call a module `warn` or similar). It can lead to hard to diagnose bugs.

**Comment by fkiraly** [comment]
>  But if the algorithms are grouped together I would see the following recurring question in our future:
> "Why do you have the same algorithm multiple times?"

I think this is a question that we *want* to get, because then we can answer:

"there are many popular packages which implement a single algorithm - for instance `optuna` implements a small family of sampling/pruning search algorithms. We give users a clear overview of where they can get the algorithm they are looking for, and a clear choice of multiple popular backends. By the way, we think the `gfo` implementation is best in cases A, B, C."

**Comment by fkiraly** [comment]
For practicalities, I will proceed with your suggested taxonomy for now, and use `sk` for `sklearn`-adjacent designs, so people will not be confused about source package - is that a good solution?

> Could you provide an example how your suggested import structure could look like?

Sure! On the top level, we either cluster individual algorithms, or close families of algorithms.
With current examples, it could be:

* `gridsearch`
    * `_gridsearch_gfo.py`
    * `_gridsearch_optuna.py`
    * `_gridsearch_sk.py`
* `hillclimbing`
    * `_hillclimbing_gfo.py`
    * `_hillclimbing_stochastic_gfo.py`

Now, I realize that where this system breaks is with "bundle" interfaces like `optuna`, where different choices of samplers and pruners allow to produce different gfo algorithms such as grid search and random search under a single interface. I see two ways to deal with this:

* `optuna` folder for the bundle
* `sample_prune` for the family of algorithms, and currently `_sp_optuna.py` the only member.

**Comment by SimonBlanke** [comment]
Based on this folder structure, how would the import look like? 

This way we have a redundant "grid_search" in the import:
```python
from hyperactive.opt.grid_search import GfoGrid_search
```

We also cannot do this:
```python
from hyperactive.opt.grid_search import Gfo
```

Do you have a suggestion how to do this?
If the optimizers are grouped by packages we could avoid this problem.

**Comment by SimonBlanke** [comment]
> No matter how we group them, I think we should give classes unique names 

Agreed

If we order by algorithm, this would be my preference:
```python
# in the "order by algorithm" case
from hyperactive.opt.grid_search  import GridSearchSk, GridSearchGfo
```
We have the algo name two times in the import, which is not so clean. But it is better, that putting all (>30 ?) algos into hyperactive.opt.

> optionally, the "best" grid search considered by us

Let's not do this. At least not for now. There should not be a "best", because there isn't mathematically. And packages or implementations can vary over time, which might shift what the best would be.

**Comment by fkiraly** [comment]
so which ordering do we go with now?

**Comment by SimonBlanke** [comment]
We order by algorithm, as you suggested. The package abbreviations make the class names short and the classes are unique. This should work out fine, agreed?

**Comment by fkiraly** [comment]
ok - I will make the changes in https://github.com/SimonBlanke/Hyperactive/pull/121 then

**Comment by SimonBlanke** [comment]
V5 is released and a solution was found.
